pkg/broker/broker.go

```diff
@@ -0,0 +1,594 @@
+package broker
+
+import (
+	"context"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"encoding/json"
+	"encoding/pem"
+	"errors"
+	"strconv"
+	"time"
+
+	"code.gitea.io/sdk/gitea"
+	"github.com/pivotal-cf/brokerapi"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+
+	"github.com/presslabs/gitea-service-broker/pkg/cmd/options"
+)
+
+// nolint: golint
+var (
+	ProvisioningStatus   = "provisioning"
+	ProvisionedStatus    = "provisioned"
+	DeprovisioningStatus = "deprovisioning"
+	BindingStatus        = "binding"
+	BoundStatus          = "bound"
+	UnbindingStatus      = "unbinding"
+)
+
+type Binding struct {
+	HostURL   string
+	CloneURL  string
+	PublicKey string
+}
+
+type ProvisionParameters struct {
+	RepositoryName         string `json:"repository_name"`
+	RepositoryOrganization string `json:"repository_organization"`
+	RepositoryOwner        string `json:"repository_owner"`
+	MigrateURL             string `json:"migrate_url"`
+}
+
+type BindingParameters struct {
+	Title     string `json:"title"`
+	PublicKey string `json:"public_key"`
+}
+
+type GiteaServiceBroker struct {
+	client.Client
+	GiteaClient
+}
+
+func (giteaServiceBroker *GiteaServiceBroker) Services(ctx context.Context) ([]brokerapi.Service, error) {
+	return []brokerapi.Service{
+		{
+			ID:          options.ServiceID,
+			Name:        options.ServiceName,
+			Description: "A Gitea service to provide repositories",
+			Bindable:    true,
+			Plans:       planList,
+			Metadata: &brokerapi.ServiceMetadata{
+				DisplayName: "Gitea",
+			},
+			Tags: []string{
+				"presslabs",
+				"gitea",
+			},
+		},
+	}, nil
+}
+
+func (giteaServiceBroker *GiteaServiceBroker) Provision(ctx context.Context, instanceID string, provisionDetails brokerapi.ProvisionDetails, asyncAllowed bool) (spec brokerapi.ProvisionedServiceSpec, err error) {
+	spec = brokerapi.ProvisionedServiceSpec{}
+
+	// Get and validate the provision parameters
+	parameters := ProvisionParameters{}
+	err = json.Unmarshal(provisionDetails.GetRawParameters(), &parameters)
+	if err != nil {
+		return spec, err
+	}
+
+	if parameters.RepositoryName == "" {
+		return spec, errors.New("`repository_name` must be specified")
```

Comment:
Extract all erros as vars and use `brokerapi.NewFailureResponse` to return meaningfull status codes.